File: examples/manual-wx-menus.rb
#! /usr/bin/env ruby
%w( rubygems wx ).each {|lib| require lib }

# To discover my highlight event bug, I need to create a very
# simple Icon with an item and a submenu and get the different
# events working properly - test this OK on Windows / OS X - 
# then fix SimpleTray
#
# I intend on implementing EVERY POSSIBLE EVENT that I might want 
# to take advantage of here, and then implementing them in SimpleTray

class AnItem < Wx::MenuItem
  def initialize *args
    super *args

    @menu = args.first # parent menu, which this item is inside of ... or whatever

    # @menu.evt_menu_highlight(self){  }

    @submenu = Wx::Menu.new #'AnItem @menu'
    @submenu.append 'First Item'
    @submenu.append 'Second Item'
    @submenu.evt_menu_close { puts "@submenu closed!" }
    @submenu.evt_menu_open { puts "@submenu opened!" }
    @submenu.evt_menu_highlight_all{ puts "highlight ALL" }

    self.set_sub_menu @submenu
  end
end

class MyTaskBarIcon < Wx::TaskBarIcon

  def initialize
    super()
    @icon = 'my_cool_app.png'
    set_icon Wx::Icon.new( @icon, Wx::BITMAP_TYPE_PNG ), 'w00t'
  end

  def create_popup_menu
    menu = Wx::Menu.new 'Hello There'
    menu.append 'I do Nothing'
    clicky = menu.append 'Clicky'
    menu.evt_menu(clicky) { puts "Clicky!!!" }
    #menu.append_item MembersMenuItem.new(@menu, 7000, 'Members')
		menu.append_separator
    
    an_item = AnItem.new( menu, 12345, 'AnItem.new' )
    menu.append_item an_item
    menu.evt_menu(an_item){ puts "something having to do with an_item!" }
    menu.evt_menu_highlight(an_item){ puts "highlighted an_item!" }
    menu.evt_menu_highlight_all{ puts "highlight ALL" }
    menu.evt_menu_close { puts "top-level menu closed!" }
    menu.evt_menu_open { puts "top-level menu opened!" }

		menu.append_separator
		quit_item = menu.append Wx::ID_ABOUT, 'About'
		quit_item = menu.append Wx::ID_EXIT, 'Quit'
    menu.evt_menu(quit_item) { exit }
    menu
  end

end

puts "Starting ..."
Wx::App.run { MyTaskBarIcon.new }

File: examples/manual-wx-menus.py
#! /usr/bin/env python
import wx

class App(wx.App):
	def OnInit(self):
		frame = wx.Frame(parent=None, title='Hello from wxPython App')
		frame.Show()
		self.tbIcon = MyIcon(self)
		return True

	def showLogin():
		print('show login!');

	def onExit(self,event):
		self.Destroy()
		self.tbIcon.Destroy()
		raise SystemExit # or import sys + call sys.exit()

class AnItem(wx.MenuItem):
	def __init__(self, parent, *args, **kwargs):
		print("hello from AnItem __init__")
		wx.MenuItem.__init__(self, parent, *args, **kwargs)
		self.parent = parent
		submenu = wx.Menu()
		submenu.Append(wx.NewId(), 'Hello')
		submenu.Append(wx.NewId(), 'There')
		self.SetSubMenu(submenu)
		self.parent.Bind(wx.EVT_MENU_HIGHLIGHT, self.OnMenuHighlight)

	def OnMenuHighlight(self,event):
		print('hello from OnMenuHighlight in AnItem')

	def __onWxMenuHighlight(self,event):
		print("haha!  overriding default __onWxMenuHighlight ... muhahahaha!")

class MyIcon(wx.TaskBarIcon):
	ID_LOGIN = wx.NewId()
	ID_CONFIG = wx.NewId()
	ID_MUTE = wx.NewId()

	def __init__(self, parent, *args, **kwargs):
		wx.TaskBarIcon.__init__(self, *args, **kwargs)
		self.parent = parent
		icon = wx.Icon("my_cool_app.png", wx.BITMAP_TYPE_PNG)
		self.SetIcon(icon, "Hello from wxPython Icon")

	def showMenu(self, event):
		self.PopupMenu(self.menu)

	def CreatePopupMenu(self):
		self.Bind(wx.EVT_TASKBAR_RIGHT_UP, self.showMenu)
		self.menu = wx.Menu()
		self.menu.Append(MyIcon.ID_LOGIN, "&Login/Change user")
		self.Bind(wx.EVT_MENU, self.parent.showLogin, id=MyIcon.ID_LOGIN)
		self.menu.AppendSeparator()
		self.menu.Append(wx.ID_EXIT, "E&xit")

		new_id = 12345
		new = AnItem(self.menu, new_id, '&New\tCtrl+N', 'Creates a new document')
		self.menu.AppendItem(new)
    		# menu.evt_menu_highlight(an_item){ puts "highlighted an_item!" } # <--- ruby
		# self.Bind(wx.EVT_MENU_HIGHLIGHT, ...
		self.Bind(wx.EVT_MENU_HIGHLIGHT, self.OnHighlightAnItem) #, id=new_id)

		#self.menu.AppendItem(wx.NewId(), AnItem(self, wx.NewId(), 'SomeTextForAnItem__init__'))
		self.Bind(wx.EVT_MENU, self.parent.onExit, id=wx.ID_EXIT)
		self.Bind(wx.EVT_MENU_HIGHLIGHT, self.onHighlightExit, id=wx.ID_EXIT)
		self.Bind(wx.EVT_CLOSE, self.parent.onExit)
		return self.menu

	def OnHighlightAnItem(self,event):
		print("OnHighlightAnItem!!!!!")

	def onHighlightExit(self,event):
		print("highlighted EXIT")

print "starting ..."
app = App()
app.MainLoop()
